### DISCUSSION TITLE
JIT not working properly in v3 upgrade

### DISCUSSION THREAD

**Discussion by fredbouin** [tailwindlabs/tailwindcss#6754] [Help] [answered]
I've just upgraded to v3 and with JIT mode I have to "npm run dev" with each new css change in order to see anything on my browser.

**Comment by adamwathan** [comment]
Hey! You have to provide much more information than this for anyone to help. Please create a public GitHub repo that reproduces the problem so someone can figure out what the issue is.

**Reply by fourwhitesocks** [reply]
Hello there, thankful for any help!  I'm having the same issue with a sveltekit app after upgrading to TW v3.  I followed the TW upgrade guide...?  Here is a repo for demonstration :) 

(https://github.com/fourwhitesocks/mattsmarinetailwind3)

**Reply by fourwhitesocks** [reply]
- Maybe we need a 'watch' script OR Vite needs updating?  
Thank you all for your time @adamwathan  and @fredbouin !!

**Comment by fredbouin** [comment]
My site is a Laravel app built with the Statamic CMS. Statamic provides tailwind by default with any new installation. I just noticed that I'm seeing the problem not just with my current site which I am upgrading to Tailwind V3 but also if I install a brand new Statamic project which comes stock with tailwind v3. Here is a repo with a new statamic install: https://github.com/fredbouin/twv3.git

On the layout.antlers.html page I have a few classes added to the <p> tag which currently don't work, but will work after compiling with "npm run dev". So far it seems as though after having compiled a class once it starts working with no hangups.

**Reply by fourwhitesocks** [reply]
That is exactly how mine reacts... it all works wonderful; except I have to stop and re-start the server ~ npm run dev.  When I have a few more monents I will check out your repo; thank you for responding and helping add details to the issue :)  
~ Fourwhitesocks aka Amy :)

**Reply by fourwhitesocks** [reply]
** I tried updating Svelte no go :/

**Comment by fourwhitesocks** [comment]
Just wanted to add that i'm using node v14.15.1 if that makes a difference for anyone looking!

**Comment by fourwhitesocks** [comment]
@fredbouin were you ever able to solve this??

**Reply by fredbouin** [reply]
unfortunately no, I'm just living with it at the moment but it's pretty annoying.

**Reply by fourwhitesocks** [reply]
I wonder if maybe we should either delete this thread and or start a new one with this time providing all of the info they need right up front; I have a feeling that is why this isn't getting answered... let me know your thoughts and I can do it if you think we should; it's soooo frustrating having to stop and start the dev server every time :(  Thank you for answering me!!

**Reply by NovusOpus** [reply]
@fourwhitesocks Hey, I'll take a look at the repo you posted above, and hopefully finding a solution.

**Reply by fourwhitesocks** [reply]
@NovusOpus  YAY!!!!  Thank you!!

**Reply by NovusOpus** [reply]
@fourwhitesocks First I though it was a problem related to HMR not working correctly. But that was not the case, After taking a more careful look at your repo I found a few mistakes in your code, for example in your `/src/routes/index.svelte` file you have this mistake:

```html
<script>
	import Header from '$lib/header.svelte';
	import Hero from '$lib/Hero.svelte'; 
</script>
```
The `Hero` import is failing, because the name of the file is lowercase:  `hero.svelte`, So any change to that file was being detected by Vite but was not being pushed by HMR Since it wasn't correctly imported. The Right import should be: 
```html
<script>
	import Header from '$lib/header.svelte';
	import Hero from '$lib/hero.svelte'; 
</script>
```
Also there's a warning with your `services button` link: `A11y: '#' is not a valid href attributesvelte(a11y-invalid-attribute)`, you can fix it by replacing `#` with `/`. (This is not really breaking anything but it's spamming the console with every HMR Update, and also being highlighted with a warning if you use [Svelte for VS Code](https://marketplace.visualstudio.com/items?itemName=svelte.svelte-vscode))

After fixing those,  everything seems to be working fine from my side, tailwind classes are being correctly generated and the HMR is updating the changes correctly without the need to restart the dev server again.

Can you verify if that fix your problems, or if there's something else not working correctly.

**Reply by fredbouin** [reply]
@NovusOpus thanks for taking the time to look into this, @fourwhitesocks let me know if this solves your issue. I expected it to be a more general sort of problem I guess that this specific fix won't apply to me, @NovusOpus if you can take a look into my repo I would really appreciate, or point me into where I should be looking.

I do find it odd how a new Statamic install which comes with tailwind 3 already setup causes the same issue for me. I'll post this in the Statamic discussion if you don't see anything in my files. Thanks!

**Reply by NovusOpus** [reply]
@fredbouin Sure, I'll take a look later today.

**Reply by NovusOpus** [reply]
@fredbouin I'm installing and setting up Statamic and I noticed [this in the `Another Hot Tip` section](https://statamic.dev/installing/local#whats-next):

> The default install and all first-party Starter Kits use TailwindCSS in Just In Time mode, so anytime you change classes in your HTML you'll need to recompile your CSS.
> 
> This is super easy and happens automatically when you doing `npm run watch` from the terminal in your project directory (as long as you've run npm install first).

So the fix is to use `npm run watch` or `npm run hot` [for HMR(Hot Module Replacement)](https://laravel-mix.com/docs/5.0/hot-module-replacement), because it's configured that way by Statamic. The `npm run watch`/`npm run hot` are running the same command than `npm run dev`:  `mix` (npm run dev:`"dev": "mix"` , npm run watch:  `"watch": "mix watch"`, npm run hot: `"hot": "mix watch --hot"`) but also looking for changes and updating your content, so there's no reason to keep using `npm run dev`.

Also apart from that i'm seeing this in the console: Your current PostCSS version is 8.4.5, but postcss-preset-env uses 7.0.39.  The version of the `postcss-preset-env` installed is 6.7.0, So there's a conflict with 2 versions of PostCSS, and Tailwind CSS V3 needs PostCSS 8, you can update `postcss-preset-env` to version [7.0.0](https://github.com/csstools/postcss-preset-env/releases/tag/7.0.0) or later that have support for PostCSS 8.

**Reply by fredbouin** [reply]
aha thanks so much, I can't believe I missed that detail! Just tested watch and it seems to be working. I'll look into the post css version thanks for pointing that out aswell

**Reply by fourwhitesocks** [reply]
@fredbouin  I will check this out as soon as I can and report back!  THANK YOU!

**Reply by fourwhitesocks** [reply]
IT'S WORKING :)  YAY!! So wow that was really dumb on my part; easy fix!  Thank you so much!!  Feeling kinda dumb I should have seen that!   I'm a happy camper!

**Reply by fourwhitesocks** [reply]
@fredbouin So we both got it solved!  Gotta luv the help you get here!!

**Comment by NovusOpus** [comment, marked as answer]
In your case with Statamic the solution is to use: `npm run watch` or `npm run hot` for [HMR(Hot Module Replacement)](https://laravel-mix.com/docs/5.0/hot-module-replacement), they both run the same command as `npm run dev`, but they also looking for changes and updating your content when detected.

**Comment by Madd0g** [comment]
Uh oh, thanks for this, I realized that I had turned off the hmr in my vite config. This breaks CSS updates.

```js
// vite.config.cjs
server: {
    hmr: false, // stops TW from updating
},

```

This should probably be fixed, watch and HMR are different.
